### ISSUE TITLE
Implement Hash-based FunctionGraph equality checks. 

### ISSUE THREAD

**Issue by kc611** [aesara-devs/aesara#724]
Currently we don't have a way for checking `Graph`/`Op` equality unless we manually traverse through it and check every corresponding node. As a solution, a hash-based signature for every `Apply` node we construct might give us a cheap way for checking equality of `Apply` nodes and hence the `Graph` constituted by those `Apply` nodes. 

The hash can be constructed as follows: 
 - Every `TensorVariable` will have a hash signature which will be some function of it's shape. 
 - Every `TensorConstant` will have a hash signature which will be some function function of it's shape and values.
 -  Every `Apply` node will have a hash signature which will be a function of the hashes of it's inputs arranged in some (alphabetical?) order and, if any, properties which affect the output values. The `TensorVariable` representing the output will take on the same signature (or a modification of it).
 - `FunctionGraph`s can be simply checked by checking equality of hashes of all it's outputs. 
 
The drawback to such checking is that it requires the `Apply` nodes to be inherently immutable structures, otherwise any change in-place within a `Apply` node will invalidate the already constructed hashes of the nodes that depend upon it/have it as an input. As of now, mutating `Apply` nodes in-place is a common practice throughout the library. For instance, the `FunctionGraph.replace` replaces the respective `Apply` nodes in-place every time an optimization occurs. This is mitigated to some extent by https://github.com/aesara-devs/aesara/pull/666 (which by the, already does require such a system to prevent costly equality checks). Anyways, we might want to take a couple steps more in that direction. i.e. making graphs immutable if we are to go forward with this. Unless we somehow keep track of all the `Apply` nodes built by a specific `Variable` which, at a global scope, might be hard as well as wasteful. 

Another thing is, we might need to differentiate between graphs/ `Op`s with a strict equality match (i.e. Their inputs are same `Variables`, but `Apply` nodes are just different instances of same `Op`s) and graphs/ `Op`s with similar structure (i.e. Their `Apply` nodes match but have two different `Variable`s with similar properties building them). In this case we might want two hashes one which takes checks the equality of the 'pattern' of the graph, and another which checks the 'identity' of the `Variable`s constituting that graph.

**Comment by brandonwillard** [comment]
> Currently we don't have a way for checking `Graph`/`Op` equality unless we manually traverse through it and check every corresponding node.

Just so everyone knows, we have `aesara.graph.basic.equal_computations` and the somewhat redundant `aesara.graph.graph_utils.is_same_graph` functions that check for "syntactic" equality (i.e. they walk both graphs and compare elements).  The important point is that these graph comparisons are not available via `Variable.__eq__`, which they arguably should be.

Graph traversal is ultimately a necessity, but it can be mitigated by performing some form of memoization/caching, which is what ultimately leads to the need for fully functional `__eq__` and `__hash__` implementations in each core graph class (e.g. `Variable`, `Op`, `Apply`, and `Type`).

The trick is that the `__hash__` results will need to be cached; otherwise, `Variable.__hash__` will need to walk the graph and compute a hash value every single time it's hash value is accessed, which could be often.

In order to cache hash values, the underlying objects need to be effectively immutable, or some complicated updating mechanism is required.  This takes us right to https://github.com/aesara-devs/aesara/discussions/421.

We can make this issue about implementing https://github.com/aesara-devs/aesara/discussions/421, and&mdash;as a direct consequence&mdash;`__eq__` and `__hash__`.  I absolutely think we should make this major refactoring ASAP.


NB: The original efficiency concerns regarding functional `__eq__` and `__hash__` implementations are largely a moot point, because, if one wants/needs to use identity-based comparisons, `is` will always work regardless of `__eq__` and `__hash__`, and, if one wants to use "identity-based" hashing, the objects can be wrapped, or a special collection type can be used.  Plus, our use of `dict`s and `set`s isn't so extensive that the latter would be infeasible or overly cumbersome.

Likewise, an `__eq__` implementation can always make early use of `is`, so comparisons like `x == x` would be (performance-wise) equivalent to `x is x`.

In other words, we could implement very naive&mdash;albeit useful&mdash;`__eq__` and `__hash__` methods right now, but we would need to employ the aforementioned in order to keep the current level of efficiency.
